### DISCUSSION TITLE
Dynamic variables in certificates

### DISCUSSION THREAD

**Discussion by andre0ani** [chamilo/chamilo-lms#5866] [General]
Hello,

I would like to generate certificates with dynamic variables, but they are not replaced by values.
Would you have a working example please?

Thanks in advance,

Patrice

**Comment by contidos** [comment]
Hello AndreOani. I can share with you a working example. I have been using variables for taking the ID numbers, names and surnames of students in our certificates. 

Here you have an example of a working one: 

![2024-10-18_20-01-22](https://github.com/user-attachments/assets/444359dd-d201-4e05-8ea0-06e504969944)

This is the code I used: 

<!DOCTYPE html>
<html>
<head>
	<title></title>
</head>
<body>
<div style="margin-left: 50px;margin-right: 50px;">
<div>
<div style="text-align: center;margin-top: 25px;margin-bottom: 15px;"><span style="font-size:21px;">XXXXXX certifica que:</span></div>

<div style="font-size: 32px; margin-top: 30px; text-align: center;"><span style="font-size:32px;">Don/Do&ntilde;a ((user_firstname)) ((user_lastname))</span></div>

<div style="text-align: center;margin-top: 30px;"><span style="font-size:21px;">con DNI ((dni)) ha superado con &eacute;xito la acci&oacute;n formativa:</span></div>

<div style="font-size: 36px; text-transform: uppercase; margin-top: 30px; font-weight: bold; color: rgb(41, 85, 93); text-align: center;line-height: 45px;"><span style="font-size:36px;">((course_title))</span></div>

<div style="margin-top: 30px; text-align: center;"><span style="font-size:21px;"><strong>Horas lectivas:</strong> 15 h </span></div>

<div style="margin-top: 20px; text-align: center;"><span style="font-size:21px;">Expedido en XXXXX ((date_simple_certificate))</span></div>

<div style="display: flex;justify-content: space-around;align-items: center;margin-top: 20px;padding: 0 20px;">&nbsp;</div>
</div>
</div>
</body>
</html>

Hope this helps you.

**Reply by hypetraxx** [reply]
Thank you for sharing this example. It's clear and practical, and I'm sure it will be helpful for AndreOani and others working on dynamic certificate generation. Using placeholders like ((user_firstname)), ((user_lastname)), and ((course_title)) makes the process straightforward and adaptable for various contexts.

Out of curiosity, how do you handle cases where the input values might be missing or incomplete? For example, if a user hasn't provided a dni or the course_title, do you include fallback text, or does your system flag it for manual review?

Great job simplifying the process, and thanks again for sharing!

**Reply by contidos** [reply]
Hello hyperatraxx, that´s an interesting question but I have never been in that case because we force the user in the registration process to fulfill all fields so never had an empty DNI field, or course name. But as far as I know, if the system does not find data, it reports blank space.

**Comment by andre0ani** [comment]
Thank you for this example.
I must be missing something, because with my code, the variables are not replaced...
I don't know what's wrong.

Here's the complete code : 

`<!DOCTYPE html>
<html data-lt-installed="true">
<head><meta http-equiv="Content-Type" content="text/html; charset=utf-8">
	<title>{course_title} - {student_name} - {certif_date}</title>
	<meta property="og:type" content="article" /><meta property="og:title" content="{course_title} - Certificate for {student_name}" /><meta property="og:url" content="{certificate_link}" />
	<style type="text/css">{external_style}
	</style>
</head>
<body data-gramm="false" data-lt-tmp-id="lt-949265" style="background-color: rgb(255, 255, 255); cursor: auto;">
<div style="width: 1072px;margin-left: auto;margin-right: auto;">
<div style="background: url('/main/img/paper.png') repeat;padding: 10px;">
<table cellpadding="0" cellspacing="0">
	<tbody>
		<tr>
			<td>
			<table cellpadding="0" cellspacing="0">
				<tbody>
					<tr>
						<td><img alt="Side bar" src="/main/img/side-bar.png" /></td>
						<td><img alt="Side top element a" src="/main/img/side-top-a.png" /></td>
						<td><img alt="Certificate logo" height="135" src="/main/img/logo_parcours.png" width="325" /></td>
						<td><img alt="Side top element b" src="/main/img/side-top-b.png" /></td>
					</tr>
				</tbody>
			</table>
			</td>
		</tr>
		<tr>
			<td>
			<table cellpadding="0" cellspacing="0" width="100%">
				<tbody>
					<tr>
						<td style="font-size: 18px;text-align: center;font-weight: normal;padding-bottom: .2em;">
						<p>&nbsp;</p>

						<p>Le ((certif_date)) sur ((institution_name))</p>
						</td>
					</tr>
					<tr>
						<td style="text-align: center;font-weight: bold;font-size: 42px;padding-top: .5em;padding-bottom: .5em;">CERTIFICAT</td>
					</tr>
					<tr>
						<td style="text-align: center;font-size: 18px;padding-bottom: .5em;padding-top: .5em;">Ce certificat est attribu&eacute; &agrave;</td>
					</tr>
					<tr>
						<td style="padding-bottom: .5em;padding-top: .5em;font-size: 35px;text-align: center;">((student_name))</td>
					</tr>
					<tr>
						<td style="text-align: center;font-size: 18px;padding-bottom: .5em;padding-top: .5em;">pour avoir compl&eacute;t&eacute; avec succ&egrave;s</td>
					</tr>
					<tr>
						<td style="padding-bottom: .5em;padding-top: .5em;font-size: 30px;text-align: center;">((course_title))</td>
					</tr>
					<tr>
						<td style="text-align: center;font-size: 18px;padding-bottom: .5em;padding-top: .5em;">le <span>((date_simple_certificate))</span></td>
					</tr>
					<tr>
						<td>
						<table align="center" cellpadding="0" cellspacing="0">
							<tbody>
								<tr>
									<td>&nbsp;</td>
									<td><img alt="blank" src="/main/img/blank.png" /></td>
									<td>&nbsp;</td>
									<td>&nbsp;</td>
									<td><img alt="blank" src="/main/img/blank.png" /></td>
									<td><img alt="certified" src="/main/img/certified.png" /></td>
								</tr>
							</tbody>
						</table>
						</td>
					</tr>
				</tbody>
			</table>
			</td>
		</tr>
	</tbody>
</table>
</div>
</div>
</body>
</html>
`

**Comment by contidos** [comment]
You have to use the exact english terms for the variables. Check that´s teh case. And also make sure you are using (( )) In your code instead of [ . Let´s see if that helps.

**Comment by andre0ani** [comment]
I've made a very basic certificate, with (( and )) and the variables in your example :
<html data-lt-installed="true">
<head>
	<title></title>
</head>
<body data-gramm="false" data-lt-tmp-id="lt-964248">
<div>Congratulations! You have finished this learning path</div>

<div>&nbsp;</div>

<div><span>((user_firstname)) ((user_lastname))</span></div>

<div>&nbsp;</div>

<div><span>((date_simple_certificate))</span></div>
<style type="text/css">body{}
</style>
</body>
</html>

If I return in a completed course, i've don't have the good informations.

**Reply by contidos** [reply]
What do you get? Blank spaces instead? Can you share an access to your portal as teacher in that course for me to take a look and see if I can find the reason why is not working? Send it to noabacu @ hotmail . com and let me know here.
